### PR TITLE
Logging to Excel file when an item is locked.

### PR DESCRIPTION
**Pull request by bmiller-jama**
Logging to Excel file when an item is locked.

### CODE REVIEW

**Comment by bmiller-jama** [comment]
@nmchale-jama I got the script to add multiple rows for one locked item and cleaned up the other things we discussed.  Let me know if you have any questions!

File: link_fixer.py [hunk 1]
```diff
@@ -276,6 +303,14 @@ def get_synced_item(item_id, project_id):
     for item in items:
         item_id = item.get('id')
         fields = item.get('fields')
+        name_field_for_excel = item['fields']['name']
+        link_in_description = item['fields']['description']
+
+        # Getting lock properties, so we don't need to call the API multiple times for Excel logging
+        item_lock_properties = (item.get('lock'))
```
Comment (nmchale-jama) [inline link_fixer.py]: remove outer parens

File: link_fixer.py [hunk 2]
```diff
@@ -440,21 +476,29 @@ def get_synced_item(item_id, project_id):
...
+                            'counter': str(bad_link_count)
+                        }
+                        broken_list = broken_link_map.get(item_id)
+                        if broken_list is None:
+                            broken_list = [broken_link_data]
+                        else:
+                            broken_list.append(broken_link_data)
+                        broken_link_map[item_id] = broken_list
+
+    workbook.save("locked_items.xlsx")
```
Comment (nmchale-jama) [inline link_fixer.py]: Move save workbook to after post phase.

Check any failed patch response for the word locked in the reason.  and add to workbook.

Comment (nmchale-jama) [reply link_fixer.py]: may need to add "item Name" to "broken_link_data" for access in post phase

**Review by nmchale-jama — CHANGES_REQUESTED** [review]

**Comment by bmiller-jama** [comment]
@nmchale-jama Made the changes we discussed on the Atari Link Fixer script: "Removed tuple for Excel headers, added functionality to check for locked items after script starts, and broken link map adjustments/logging for that."

File: link_fixer.py [hunk 3]
```diff
@@ -208,6 +210,29 @@ def get_synced_item(item_id, project_id):
...
+    return workbook
+
+
+def log_locked_items(workbook, item_name, item_id, key, hyperlink):
+    sheet = workbook.active
+    # Check for item lock
+    if client.get_item_lock(item_id):
+        row = [item_name, item_id, key, hyperlink]
+        sheet.append(row)
+        logger.info("Item {} is locked and was added to the Excel workbook.")
```
Comment (cmarqu) [inline link_fixer.py]:
```suggestion
        logger.info("Item {} is locked and was added to the Excel workbook.".format(item_name))
```

File: link_fixer.py [hunk 4]
```diff
@@ -440,21 +476,29 @@ def get_synced_item(item_id, project_id):
...
-                if broken_list is None:
-                    broken_list = [broken_link_data]
-                else:
-                    broken_list.append(broken_link_data)
-                broken_link_map[item_id] = broken_list
+                # we have a bad link for this item?
+                if bad_link_found:
+                    # Before we replace the hyperlink, let's check if it's locked and log it to Excel if so
+                    if item_lock_properties['locked']:
+                        print("Item was locked and has a broken link.  Logging to Excel File...\n")
```
Comment (trobinson-jama) [inline link_fixer.py]: Change to logger.warn statement

File: link_fixer.py [hunk 5]
```diff
@@ -489,19 +533,29 @@ def get_synced_item(item_id, project_id):
...
+                    logger.info('Successfully patched item [' + b.get('name') + ']')
                 except APIException as error:
-                    # failed to patch
-                    logger.error('Failed to patched item [' + str(item_id) + ']')
-                    logger.error('API exception response: ' + str(error))
+                    if "locked" in str(error):
+                        try:
+                            log_locked_items(workbook, str(b.get('name')), str(b.get('itemId')), str(b.get('fieldName')),
+                                             str(logger_old_value))
+                            print("Log locked items method successful for " + str(b.get('name')))
```
Comment (trobinson-jama) [inline link_fixer.py]: Change to logger.info statement, and all other print statement

File: requirements.txt
Comment (trobinson-jama) [inline requirements.txt:1]: Change py rest client to 17.1

**Review by trobinson-jama — APPROVED** [review]

**Pull request merged**